active/0020-node-evacuation.md

```diff
@@ -0,0 +1,399 @@
+* Enables/Disables Eviction Agent on nodes.
+* Sends connection/session eviction commands according to the evacuation logic.
+
+Rebalance Coordinators contain logic for active connection eviction but know little about connection/session nature.
+
+This component decomposition is chosen to decouple working with low-level details from the high-level logic of evacuation:
+* Eviction Agent is aware of connection handling details (hooks, connections, sessions, connection managers, ...);
+* Rebalance Coordinators know the general algorithm and handle connection arithmetics.
+
+That should allow us to reuse Eviction Agents in different scenarios, like node evacuation and node rebalancing.
+
+## Components
+
+### Virtual Connection pool
+
+To preserve sessions for clients that do not reconnect, we implement a _virtual_
+connection. It is a simplified version of `emqx_connection`/`emqx_ws_connection`
+that does not have a real associated socket or WS connection.
+
+These virtual connections do not run under `esockd` or `ranch` supervisors but under
+their own supervisor.
+
+### Eviction Agent
+
+Eviction Agent is a stateful process and a collection of supporting modules. Eviction Agent is
+passive and responsible for disconnecting clients.
+It includes:
+* Stateful process that handles commands to manage eviction state.
+* `connect` and `connack` hook logic.
+* Facade internal API for
+    * connection, session eviction;
+    * reporting current connection & session stats.
+* HTTP API reporting current status and providing endpoints for health checks.
+
+Eviction Agent has two states: `enabled` and `disabled`.
+
+In the `enabled` state, Eviction Agent has _kind_, a label that indicates the purpose for which the agent is used.
+This label is used to avoid a clash between different Rebalance Coordinators, like evacuation and rebalancing.
+
+#### Internal API
+
+Eviction Agent exposes API for the following actions:
+* In disabled state:
+    * enable the agent;
+* In either of enabled states:
+    * disable the agent;
+    * provide stats about connected/disconnected client count;
+    * provide stats about session count;
+    * disconnect certain amount of connected clients;
+    * evict certain amount of sessions.
+
+#### HTTP API
+
+Eviction Agent exposes two endpoints: `/node_eviction/status` and `/node_eviction/heath_check`.
+
+#### Algorithm
+
+##### `disabled` state
+
+The disabled state is the default. In the `disabled` state Eviction Agent does not influence the node anyhow.
+
+HTTP responses are the following:
+
+Request:
+```
+GET /api/v4/node_eviction/status
+```
+
+Response:
+```
+HTTP/1.1 200 OK
+
+{"status": "disabled"}
+```
+
+Request:
+```
+GET /api/v4/node_eviction/heath_check
+```
+
+Response:
+```
+HTTP/1.1 200 OK
+```
+
+The internal API behavior is the following:
+
+```erlang
+disabled = emqx_eviction_agent:status().
+```
+
+##### Enable
+
+Enabling Eviction Agent requires two arguments:
+* `kind` label.
```

Comment:
Q: can it be anything, or is it an enumeration of choices?

Reply:
Any atom that makes sence only for the client of the API.